x-pack/platform/plugins/shared/fleet/public/applications/fleet/sections/agents/agent_list_page/hooks/use_session_agent_list_state.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback } from 'react';
+import type { CriteriaWithPagination } from '@elastic/eui';
+import useSessionStorage from 'react-use/lib/useSessionStorage';
+
+import type { Agent } from '../../../../types';
+
+const FLEET_NAMESPACE = 'fleet';
+const AGENT_LIST_SESSION_STORAGE_KEY = 'agentListState';
+
+interface AgentListTableState {
+  search: string;
+  selectedAgentPolicies: string[];
+  selectedStatus: string[];
+  selectedTags: string[];
+  showUpgradeable: boolean;
+  sort: {
+    field: keyof Agent;
+    direction: 'asc' | 'desc';
+  };
+  page: {
+    index: number;
+    size: number;
+  };
+}
+
+type SessionAgentListState = AgentListTableState & {
+  clearFilters: () => void;
+  updateTableState: (partialState: Partial<AgentListTableState>) => void;
+  onTableChange: (changes: Partial<CriteriaWithPagination<Agent>>) => void;
+};
+
+export const defaultAgentListState: AgentListTableState = {
+  search: '',
+  selectedAgentPolicies: [],
+  selectedStatus: ['healthy', 'unhealthy', 'orphaned', 'updating', 'offline'],
+  selectedTags: [],
+  showUpgradeable: false,
+  sort: {
+    field: 'enrolled_at',
+    direction: 'desc',
+  },
+  page: {
+    index: 0,
+    size: 20,
+  },
+};
+
+export const useSessionAgentListState = (): SessionAgentListState => {
+  const fullStorageKey = `${FLEET_NAMESPACE}.${AGENT_LIST_SESSION_STORAGE_KEY}`;
+
+  const [sessionState, setSessionState] = useSessionStorage<AgentListTableState>(
+    fullStorageKey,
+    defaultAgentListState
+  );
+
+  // Utility functions
+  const updateTableState = useCallback(
+    (partialState: Partial<AgentListTableState>) => {
+      const newState = { ...sessionState, ...partialState };
+      setSessionState(newState);
+    },
+    [sessionState, setSessionState]
+  );
+
+  // Atomic update function for table changes - prevents multiple rapid updates
+  const onTableChange = useCallback(
+    (changes: Partial<CriteriaWithPagination<Agent>>) => {
+      const latestState = sessionState;
+      const updates: Partial<AgentListTableState> = {};
+
+      // Check if pagination has actually changed
+      if (
```

Comment:
nit: For the comparison checks between the updates and current state, I wonder if we could lean on lodash [isMatch](https://lodash.com/docs/4.17.15#isMatch) for a partial deep compare

Reply:
changed